databricks_cli/stack/api.py

```diff
@@ -0,0 +1,340 @@
+
+        :param stack_path: Path to the stack config template JSON file
+        :return: The path to the stack status file.
+
+        >>> self._generate_stack_status_path('./stack.json')
+        './stack.deployed.json'
+        """
+        stack_path_split = stack_path.split('.')
+        stack_path_split.insert(-1, STACK_STATUS_INSERT)
+        return '.'.join(stack_path_split)
+
+    def _get_previous_resource_deploy_map(self, stack_status):
+        """
+        Returns a dictionary that maps a resource's (id, service) to the resource's metadata
+        from the last deployment
+        """
+        return {(resource[RESOURCE_ID], resource[RESOURCE_SERVICE]): resource
+                for resource in stack_status[STACK_DEPLOYED]}
+
+    def put_job(self, job_settings):
+        """
+        Given settings of the job in job_settings, create a new job. For purposes of idempotency
+        and to reduce leaked resources in alpha versions of stack deployment, if a job exists
+        with the same name, that job will be updated. If multiple jobs are found with the same name,
+        the deployment will abort.
+
+        :param job_settings:
+        :return: job_id, Physical ID of job on Databricks server.
+        """
+        if 'name' not in job_settings:
+            raise StackError("Please supply 'name' in job resource 'resource_properties'")
+        job_name = job_settings['name']
+        jobs_same_name = self.jobs_client._list_jobs_by_name(job_name)
+        if len(jobs_same_name) > 1:
+            raise StackError("Multiple jobs with the same name '%s' already exist, aborting"
+                             " stack deployment" % job_name)
+        elif len(jobs_same_name) == 1:
+            existing_job = jobs_same_name[0]
+            creator_name = existing_job['creator_user_name']
+            timestamp = existing_job['created_time'] / MS_SEC  # Convert to readable date.
+            date_created = datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
+            click.echo("Warning: Job exists with same name '%s' created by %s on %s. Job will "
+                       "be overwritten" % (job_name, creator_name, date_created))
+            self.update_job(job_settings, existing_job['job_id'])
+            return existing_job['job_id']
+        else:
+            click.echo("Creating new job")
+            job_id = self.jobs_client.create_job(job_settings)['job_id']
+            return job_id
+
+    def update_job(self, job_settings, job_id):
+        """
+        Given job settings and an existing job_id of a job, update the job settings on databricks.
+
+        :param job_settings: job settings to update the job with.
+        :param job_id: physical job_id of job in databricks server.
+        """
+        click.echo("Updating Job")
+        self.jobs_client.reset_job({'job_id': job_id, 'new_settings': job_settings})
+
+    def deploy_job(self, resource_id, resource_properties, physical_id=None):
```

Comment:
Why does this method need to know `resource_id`

Reply:
technically doens't need to.